# bioe131
Shirley Zhou's Intro to Computational Biology repo

## About
This repo is Shirley Zhou's assignment submission of Fall 2019 BioE 131 Introductio to Computational Biology

## Assignment List

**Lab 1**: Set up server and github environment <br>
**Lab 2**: Align, identify, and build a phylogenetic tree from a set of closely related sequences; makes use of `muscle` and `fasttree` <br>
**Lab 3**: Python exercise. Create a dictionary in which keys are unique words and the values are the number of times that word appears in Romeo and Juliet <br>
**Lab 4**: Create SQL databases and associative tables with curated genes from Entrez; makes use of `Bio.Entrez`, `Bio.SeqIO`, `sqlite3` <br>
**Lab 5**: Generate riboswitch structures with given sequences and output as pictures; makes use of `RNAfold` motif<br>
**Lab 6**: Align reads (sam file) with given bacterial reference genome and sort the aligned contigs; makes use of `bowtie2` and `samtools` <br>
**Lab 7**: Introduction to compression algorithms and test out the efficiency of `gzip`, `bzip` and `pbzip2` <br>
**Lab 8 and 9**: Assemble given bacterial sequences, identify taxon, and annotate genome using RAST; makes use of `spades.py` `assembly-stats`, `rna_hmm3.py` and `bedtools`  <br>
**Lab 10**: Implement Needleman-Wunsch (NW) algorithms that gives the best-scored alignment of two given sequences<br>
**Final Project**: Assemble given salmonella sequences, identify taxon, annotate genome using RAST, and research the selection pressure on each strain of salmonella; makes use of `spades.py` `assembly-stats`, `rna_hmm3.py` and `bedtools`; final presentation: https://docs.google.com/presentation/d/1irjqjUIL_-gp7lTkJIejlaFPV8cxYjGB_MduJRUV39g/edit?usp=sharing <br>
