{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# BioE 131 Lab 4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Author: Shirley Zhou"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sqlite3\n",
    "from Bio import Entrez\n",
    "from Bio import SeqIO\n",
    "import matplotlib as plt\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "import io\n",
    "import os"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Table Design"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For this lab assignment, we focus on three pathways: glycolysis, the citrate cycle, and the pentose phosphate pathway. We choose four enzymes in each of these pathways and search their version of gene in Drosophila, E. coli, and humans."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The tables that we would like to create in our SQL database that stores these informations are:\n",
    "\n",
    "__Gene Table__: name, description, organism, and nucleotide sequence.\n",
    "\n",
    "__Pathway Table__: name and description.\n",
    "\n",
    "__Enzyme Table__: name, function, and enzyme commission (EC) number."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "conn = sqlite3.connect('lab4.db')\n",
    "c = conn.cursor()\n",
    "c.execute('''CREATE TABLE gene (id TEXT, description TEXT, organism TEXT, seq TEXT);''')\n",
    "c.execute('''CREATE TABLE pathway (name TEXT, description TEXT);''')\n",
    "c.execute('''CREATE TABLE enzyme (name TEXT, function TEXT, ec TEXT);''')\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Pathway Table"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Description of pathways are easily found on the KEGG databases."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "glycolysis_des = \"process of converting glucose into pyruvate and generating small amounts of ATP (energy) and NADH (reducing power)\"\n",
    "citrate_des = \"important aerobic pathway for the final steps of the oxidation of carbohydrates and fatty acids\"\n",
    "pentose_des = \"a process of glucose turnover that produces NADPH as reducing equivalents and pentoses as essential parts of nucleotides\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "c.execute('''INSERT INTO pathway (name, description)\n",
    "                                 VALUES (?, ?);''',\n",
    "          (\"glycolysis\", glycolysis_des))\n",
    "c.execute('''INSERT INTO pathway (name, description)\n",
    "                                 VALUES (?, ?);''',\n",
    "          (\"citrate cycle\", citrate_des))\n",
    "c.execute('''INSERT INTO pathway (name, description)\n",
    "                                 VALUES (?, ?);''',\n",
    "          (\"Pentose phosphate\", pentose_des))\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can visualize the the table using `pandas.read_sql`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>glycolysis</td>\n",
       "      <td>process of converting glucose into pyruvate an...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>citrate cycle</td>\n",
       "      <td>important aerobic pathway for the final steps ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Pentose phosphate</td>\n",
       "      <td>a process of glucose turnover that produces NA...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                name                                        description\n",
       "0         glycolysis  process of converting glucose into pyruvate an...\n",
       "1      citrate cycle  important aerobic pathway for the final steps ...\n",
       "2  Pentose phosphate  a process of glucose turnover that produces NA..."
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.read_sql('SELECT * FROM pathway', conn)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Screen through databases for Enzymes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "From the KEGG database, I screen through some genes in each pathway graph _(glycolysis, citrate cycle, pentose phosphate)_ that are likely to be conserved in each organism _(drosophila, E.coli, Homo Sapiens)_ ."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "glycolysis = [\"phosphoglucomutase\",\n",
    "              \"glucose-6-phosphate isomerase\",\n",
    "              \"enolase\",\n",
    "              \"glucokinase\"]\n",
    "\n",
    "citrate = [\"aconitate hydratase\",\n",
    "           \"pyruvate dehydrogenase\",\n",
    "           \"succinyl-coA\",\n",
    "           \"dihydrolipoamide dehydrogenase\"]\n",
    "\n",
    "pentose = [\"ribulose-phosphate 3-epimerase\", \n",
    "           \"deoxyribose-phosphate aldolase\", \n",
    "            \"hosphogluconate dehydrogenase\", \n",
    "            \"phosphofructokinase\"]\n",
    "pw = [glycolysis, citrate, pentose]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "When looking at the enzymes on the pathway maps, the enzyme commission number were labeled on the graph already."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "glycolysis_ec = [\"5.4.2.2\",\n",
    "                 \"5.3.1.9\",\n",
    "                 \"4.2.1.11\",\n",
    "                 \"2.7.1.2\"]\n",
    "\n",
    "citrate_ec = [\"4.2.1.3\",\n",
    "              \"1.2.4.1\",\n",
    "              \"6.2.1.5\",\n",
    "              \"1.8.1.4\"]\n",
    "\n",
    "pentose_ec = [\"5.1.3.1\",\n",
    "              \"4.1.2.4\",\n",
    "              \"1.1.1.44\",\n",
    "              \"2.7.1.11\"]\n",
    "\n",
    "pw_ec = [glycolysis_ec, citrate_ec, pentose_ec]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Given the enzyme commision number, we could easily extract information (like enzyme function) from KEGG using `Bio.KEGG.REST` module. The only downside about this module is that the records extracted are in the form of iotextwrapper or string. Individual pieces of information can only be extracted out using string parsing (not very efficient)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.KEGG import REST as r"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "def enzyme_funct(ec_num):\n",
    "    term = \"ec:\" + ec_num\n",
    "    result = r.kegg_get(term).read()\n",
    "    start = result.find(\"CLASS\") + 5\n",
    "    end = result.find(\"SYSNAME\")\n",
    "    function = result[start:end].replace(\"\\n\",\"\")\n",
    "    funct =\" \".join(function.split())\n",
    "    return funct"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Enzyme Table"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now that we have every piece of enzyme information, we can construct the enzyme table."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "for i in np.arange(3):\n",
    "    for j in np.arange(4):\n",
    "        name = pw[i][j]\n",
    "        ec = pw_ec[i][j]\n",
    "        function = enzyme_funct(ec)\n",
    "        c.execute('''INSERT INTO enzyme (name, function, ec)\n",
    "                                 VALUES (?, ?, ?);''',\n",
    "                 (str(name), str(function), str(ec)))\n",
    "        conn.commit()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>function</th>\n",
       "      <th>ec</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>phosphoglucomutase</td>\n",
       "      <td>Isomerases; Intramolecular transferases; Phosp...</td>\n",
       "      <td>5.4.2.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>glucose-6-phosphate isomerase</td>\n",
       "      <td>Isomerases; Intramolecular oxidoreductases; In...</td>\n",
       "      <td>5.3.1.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>enolase</td>\n",
       "      <td>Lyases; Carbon-oxygen lyases; Hydro-lyases</td>\n",
       "      <td>4.2.1.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>glucokinase</td>\n",
       "      <td>Transferases; Transferring phosphorus-containi...</td>\n",
       "      <td>2.7.1.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>aconitate hydratase</td>\n",
       "      <td>Lyases; Carbon-oxygen lyases; Hydro-lyases</td>\n",
       "      <td>4.2.1.3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                            name  \\\n",
       "0             phosphoglucomutase   \n",
       "1  glucose-6-phosphate isomerase   \n",
       "2                        enolase   \n",
       "3                    glucokinase   \n",
       "4            aconitate hydratase   \n",
       "\n",
       "                                            function        ec  \n",
       "0  Isomerases; Intramolecular transferases; Phosp...   5.4.2.2  \n",
       "1  Isomerases; Intramolecular oxidoreductases; In...   5.3.1.9  \n",
       "2         Lyases; Carbon-oxygen lyases; Hydro-lyases  4.2.1.11  \n",
       "3  Transferases; Transferring phosphorus-containi...   2.7.1.2  \n",
       "4         Lyases; Carbon-oxygen lyases; Hydro-lyases   4.2.1.3  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.read_sql('SELECT * FROM enzyme', conn).head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Gene Table"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To search for these enzymes in the nucleotide database using `Bio.Entrez`, we need the corresponding gene name of those enzymes in the three different organisms. From the `kegg_get` result, we could use string manipulation to parse out the gene names given the 3-letter abbreviation of organism name. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_name = []\n",
    "\n",
    "for i in np.arange(3):\n",
    "    for j in np.arange(4):\n",
    "        term = \"ec:\" + pw_ec[i][j]        \n",
    "        result = r.kegg_get(term).read()\n",
    "\n",
    "        eco_start = result.find(\"ECO: \") + 5\n",
    "        eco_parse = result[eco_start:]\n",
    "        eco_gene = eco_parse[eco_parse.find(\"(\")+1 :eco_parse.find(\")\")]\n",
    "        gene_name.append((\"e.coli\", eco_gene))\n",
    "    \n",
    "        dme_start = result.find(\"DME: \") + 5\n",
    "        dme_parse = result[dme_start:]\n",
    "        dme_gene = dme_parse[dme_parse.find(\"(\")+1 :dme_parse.find(\")\")]\n",
    "        gene_name.append((\"drosophila\", dme_gene))\n",
    "\n",
    "        hsa_start = result.find(\"HSA: \") + 5\n",
    "        hsa_parse = result[hsa_start:]\n",
    "        hsa_gene = hsa_parse[hsa_parse.find(\"(\")+1 :hsa_parse.find(\")\")]\n",
    "        gene_name.append((\"homo sapiens\", hsa_gene))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('e.coli', 'pgm'),\n",
       " ('drosophila', 'Pgm2b'),\n",
       " ('homo sapiens', 'PGM1'),\n",
       " ('e.coli', 'pgi'),\n",
       " ('drosophila', 'Pgi'),\n",
       " ('homo sapiens', 'GPI'),\n",
       " ('e.coli', 'eno'),\n",
       " ('drosophila', 'Eno'),\n",
       " ('homo sapiens', 'ENO1'),\n",
       " ('e.coli', 'glk'),\n",
       " ('drosophila', 'phosphorylating'),\n",
       " ('homo sapiens', 'GCK'),\n",
       " ('e.coli', 'acnB'),\n",
       " ('drosophila', 'mAcon2'),\n",
       " ('homo sapiens', 'ACO1'),\n",
       " ('e.coli', 'aceE'),\n",
       " ('drosophila', 'Pdhb'),\n",
       " ('homo sapiens', 'PDHA1'),\n",
       " ('e.coli', 'sucC'),\n",
       " ('drosophila', 'ScsbetaG'),\n",
       " ('homo sapiens', 'SUCLG2'),\n",
       " ('e.coli', 'lpd'),\n",
       " ('drosophila', 'CG7430'),\n",
       " ('homo sapiens', 'DLD'),\n",
       " ('e.coli', 'rpe'),\n",
       " ('drosophila', 'Rpe'),\n",
       " ('homo sapiens', 'RPE'),\n",
       " ('e.coli', 'deoC'),\n",
       " ('drosophila', 'Dera'),\n",
       " ('homo sapiens', 'DERA'),\n",
       " ('e.coli', 'gnd'),\n",
       " ('drosophila', 'Pgd'),\n",
       " ('homo sapiens', 'PGD'),\n",
       " ('e.coli', 'pfkB'),\n",
       " ('drosophila', 'Pfk'),\n",
       " ('homo sapiens', 'PFKL')]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_name"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we can define a function that takes the organism and gene name and do the entrez search for us, and return the search result as handle (genbank format) that can be later read using `SeqIO`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "def entrez_search(tup):\n",
    "    org = tup[0]\n",
    "    gene = tup[1]\n",
    "    Entrez.email = 'shirleyzhou428@berkeley.edu'\n",
    "    search_term = org + '[ORGN] ' + gene\n",
    "    print(search_term)\n",
    "    handle = Entrez.esearch(db = \"nucleotide\",\n",
    "                            term = search_term, \n",
    "                            sort = 'relevance',\n",
    "                            idtype = 'acc',\n",
    "                            retmax = 1)\n",
    "        \n",
    "    for entry in Entrez.read(handle)['IdList']:  \n",
    "        result = Entrez.efetch(db='nucleotide', id=entry, rettype='gb', retmode = 'text')\n",
    "        \n",
    "        record = SeqIO.read(result, \"gb\")\n",
    "        gene_id = record.name\n",
    "        descript = record.description\n",
    "        seq = record.seq\n",
    "        \n",
    "        return (gene_id, descript, org, seq)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can define another function that handle the reading of each record, extract different informations and put them into the gene table."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "def add_to_sql_gene(tup):\n",
    "    gene_id = tup[0]\n",
    "    descript = tup[1]\n",
    "    org = tup[2]\n",
    "    seq = tup[3]\n",
    "    \n",
    "    c.execute('''INSERT INTO gene (id, description, organism, seq)\n",
    "                 VALUES (?, ?, ?, ?);''',\n",
    "              (str(gene_id), str(descript), str(org), str(seq)))\n",
    "    conn.commit()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "e.coli[ORGN] pgm\n",
      "drosophila[ORGN] Pgm2b\n",
      "homo sapiens[ORGN] PGM1\n",
      "e.coli[ORGN] pgi\n",
      "drosophila[ORGN] Pgi\n",
      "homo sapiens[ORGN] GPI\n",
      "e.coli[ORGN] eno\n",
      "drosophila[ORGN] Eno\n",
      "homo sapiens[ORGN] ENO1\n",
      "e.coli[ORGN] glk\n",
      "drosophila[ORGN] phosphorylating\n",
      "homo sapiens[ORGN] GCK\n",
      "e.coli[ORGN] acnB\n",
      "drosophila[ORGN] mAcon2\n",
      "homo sapiens[ORGN] ACO1\n",
      "e.coli[ORGN] aceE\n",
      "drosophila[ORGN] Pdhb\n",
      "homo sapiens[ORGN] PDHA1\n",
      "e.coli[ORGN] sucC\n",
      "drosophila[ORGN] ScsbetaG\n",
      "homo sapiens[ORGN] SUCLG2\n",
      "e.coli[ORGN] lpd\n",
      "drosophila[ORGN] CG7430\n",
      "homo sapiens[ORGN] DLD\n",
      "e.coli[ORGN] rpe\n",
      "drosophila[ORGN] Rpe\n",
      "homo sapiens[ORGN] RPE\n",
      "e.coli[ORGN] deoC\n",
      "drosophila[ORGN] Dera\n",
      "homo sapiens[ORGN] DERA\n",
      "e.coli[ORGN] gnd\n",
      "drosophila[ORGN] Pgd\n",
      "homo sapiens[ORGN] PGD\n",
      "e.coli[ORGN] pfkB\n",
      "drosophila[ORGN] Pfk\n",
      "homo sapiens[ORGN] PFKL\n"
     ]
    }
   ],
   "source": [
    "gene_list = []\n",
    "for item in gene_name:\n",
    "    tup = entrez_search(item)\n",
    "    gene_list.append(tup)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "for item in gene_list:\n",
    "    add_to_sql_gene(item)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>description</th>\n",
       "      <th>organism</th>\n",
       "      <th>seq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>FJ217664</td>\n",
       "      <td>Escherichia coli strain 33 Pgm (pgm) gene, par...</td>\n",
       "      <td>e.coli</td>\n",
       "      <td>ATGGTGGCCCGGCTGATACCAACGTCACTAAAGTGGTGGAAGACAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NM_137148</td>\n",
       "      <td>Drosophila melanogaster phosphoglucomutase 2b ...</td>\n",
       "      <td>drosophila</td>\n",
       "      <td>CTGATATTCAAGTGTAAGTTTACTAACCAAATTTCTAGTTTGTCAC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HUMPGM1A</td>\n",
       "      <td>Human phosphoglucomutase 1 (PGM1) mRNA, comple...</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>GGGCCGGCCGCCCCTCCGCCAGCCAAGTCCGCCGCTCTGACCCCCG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AY616650</td>\n",
       "      <td>Escherichia coli strain LW1655F+ Pgi (pgi) gen...</td>\n",
       "      <td>e.coli</td>\n",
       "      <td>GCTGCCTGGCAGGCACTACAGAAACACTTCGATGAAATGAAAGACG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>DQ363252</td>\n",
       "      <td>Drosophila jambulina Pgi (Pgi) gene, partial cds</td>\n",
       "      <td>drosophila</td>\n",
       "      <td>ATTGACAGCGCCAATATGTTCGGCTTCTGGGACTGGGTTGGCGGAC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>NM_001289789</td>\n",
       "      <td>Homo sapiens glucose-6-phosphate isomerase (GP...</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>AATAGCCCTTACCACCAGCAGACACACATCATCTGTTGTACTTGCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>MK461929</td>\n",
       "      <td>Escherichia coli strain 2009_36 plasmid p2009_...</td>\n",
       "      <td>e.coli</td>\n",
       "      <td>ATTCAGACATCAAAAAACTGTTCGGCGAGGTGGATAAGTCCTCCGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>XM_001356889</td>\n",
       "      <td>Drosophila pseudoobscura pseudoobscura Eno (Dp...</td>\n",
       "      <td>drosophila</td>\n",
       "      <td>ATGGGGTGGACCACGGATCAGGAACACACACACCTGCTCATCTCCG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>KJ896766</td>\n",
       "      <td>Synthetic construct Homo sapiens clone ccsbBro...</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>GTTCGTTGCAACAAATTGATGAGCAATGCTTTTTTATAATGCCAAC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>AFAS01000074</td>\n",
       "      <td>Escherichia coli PCN009 contig107, whole genom...</td>\n",
       "      <td>e.coli</td>\n",
       "      <td>AGCCGAAGGACATTACCTGGTCATAACGCGGACGCGGTAACGACGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>CK135585</td>\n",
       "      <td>RH55882.3prime RH Drosophila melanogaster norm...</td>\n",
       "      <td>drosophila</td>\n",
       "      <td>AAATATTTAAATGCTTATGAGTCGGCATTTTTAAAACTACACACAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>KJ905202</td>\n",
       "      <td>Synthetic construct Homo sapiens clone ccsbBro...</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>GTTCGTTGCAACAAATTGATGAGCAATGCTTTTTTATAATGCCAAC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>AFST01000003</td>\n",
       "      <td>Escherichia coli O104:H4 str. C227-11 C227-11_...</td>\n",
       "      <td>e.coli</td>\n",
       "      <td>CCCGATGGTTCTAGGCCCCGTTAGCATCCGTTGCCGCGGCAGGCCG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>NM_141785</td>\n",
       "      <td>Drosophila melanogaster mitochondrial aconitas...</td>\n",
       "      <td>drosophila</td>\n",
       "      <td>CTTCGACTTCAACAACACGAACTGCTAACCAATAAATCAACAAAAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>KJ890616</td>\n",
       "      <td>Synthetic construct Homo sapiens clone ccsbBro...</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>GTTCGTTGCAACAAATTGATGAGCAATGCTTTTTTATAATGCCAAC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>S67363</td>\n",
       "      <td>pdhR=genA, aceE=AceE [Escherichia coli, K-12, ...</td>\n",
       "      <td>e.coli</td>\n",
       "      <td>ATGCACCCGCTTTATGCATGGTTGAAGATGAGTTGCTTAAAAAGAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>NM_143411</td>\n",
       "      <td>Drosophila melanogaster pyruvate dehydrogenase...</td>\n",
       "      <td>drosophila</td>\n",
       "      <td>GCCGGCAACGATAGCGATAGACCCATAGGAAAGCGAACCGCAAACC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>EF576990</td>\n",
       "      <td>Homo sapiens PDHA1 (PDHA1) mRNA, partial cds</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>TTCCGTCCGAGAAATTCTCGCAGAGCTTACAGGACGAAAAGGAGGT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>CP015085</td>\n",
       "      <td>Escherichia coli O25b:H4 chromosome, complete ...</td>\n",
       "      <td>e.coli</td>\n",
       "      <td>CTCCACGGAACTGGTGGAAGACCCGGAAGCCATACTGCGCTACGGA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>NM_001300012</td>\n",
       "      <td>Drosophila melanogaster Succinyl-coenzyme A sy...</td>\n",
       "      <td>drosophila</td>\n",
       "      <td>CTTCTTGGTCACATTGTGCAGTGATAAGCATCGTGGAACGGCGGCG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>KJ901911</td>\n",
       "      <td>Synthetic construct Homo sapiens clone ccsbBro...</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>GTTCGTTGCAACAAATTGATGAGCAATGCTTTTTTATAATGCCAAC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>KU997026</td>\n",
       "      <td>Escherichia coli strain VA292 plasmid pDGO100,...</td>\n",
       "      <td>e.coli</td>\n",
       "      <td>GATAGGCTCAGATAAACAGACCTTACCCTCGCATCGAGAACCGCTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>LC165517</td>\n",
       "      <td>Drosophila unipectinata CG7430 gene, partial cds</td>\n",
       "      <td>drosophila</td>\n",
       "      <td>GCAGCGGCGACAACGTCACCGTGTCCGTGGAGAACGCAAAGACCGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>KR711450</td>\n",
       "      <td>Synthetic construct Homo sapiens clone CCSBHm_...</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>GTTCGTTGCAACAAATTGATGAGCAATGCTTTTTTATAATGCCAAC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>CP015085</td>\n",
       "      <td>Escherichia coli O25b:H4 chromosome, complete ...</td>\n",
       "      <td>e.coli</td>\n",
       "      <td>CTCCACGGAACTGGTGGAAGACCCGGAAGCCATACTGCGCTACGGA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>NM_001273851</td>\n",
       "      <td>Drosophila melanogaster Ribulose-5-phosphate-3...</td>\n",
       "      <td>drosophila</td>\n",
       "      <td>AGCTCCCCTCTGTGTGACCTTAAGTCAACAAGACGCCACCCGCGAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>KJ892020</td>\n",
       "      <td>Synthetic construct Homo sapiens clone ccsbBro...</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>GTTCGTTGCAACAAATTGATGAGCAATGCTTTTTTATAATGCCAAC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>KF678351</td>\n",
       "      <td>Escherichia coli strain ED205 type VI secretio...</td>\n",
       "      <td>e.coli</td>\n",
       "      <td>ATGGCTGACAGTTTCCAGAATGAAGTTCCCACCGCTCGTGTAAATA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>NM_001299419</td>\n",
       "      <td>Drosophila melanogaster Deoxyribose-phosphate ...</td>\n",
       "      <td>drosophila</td>\n",
       "      <td>TGATAAGACTCTCCATGTTCTTCGGTTGAGGTTGAGTAGAAGATTA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>NM_015954</td>\n",
       "      <td>Homo sapiens deoxyribose-phosphate aldolase (D...</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>GCGGCGCAGAGGCGGGCGCCTACCAGCCGGCAGCTCCGGAGCTGCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>DQ472577</td>\n",
       "      <td>Escherichia coli strain 02-0714 Gnd (gnd) gene...</td>\n",
       "      <td>e.coli</td>\n",
       "      <td>TTCCGGTGGTGAAGAGGGCGCGCTGAAAGGTCCTTCCATTATGCCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>CP012528</td>\n",
       "      <td>Drosophila busckii chromosome X sequence</td>\n",
       "      <td>drosophila</td>\n",
       "      <td>CACACACACACACACACACACACACACCCTGCACAACAGCGCCACT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>KJ891789</td>\n",
       "      <td>Synthetic construct Homo sapiens clone ccsbBro...</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>GTTCGTTGCAACAAATTGATGAGCAATGCTTTTTTATAATGCCAAC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>AIEW01000031</td>\n",
       "      <td>Escherichia coli DEC1B gecDEC1B.contig.30_1, w...</td>\n",
       "      <td>e.coli</td>\n",
       "      <td>CTCTGGGACCAGGAACCTCGAAAATAATAAGCGTAGTAGAATCTGT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>CP012524</td>\n",
       "      <td>Drosophila busckii chromosome 2R sequence</td>\n",
       "      <td>drosophila</td>\n",
       "      <td>GGATCTTGCCATCAGGTCCCTCGGTTGGCGTCTGTTTAGTCGCGTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>KJ905855</td>\n",
       "      <td>Synthetic construct Homo sapiens clone ccsbBro...</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>GTTCGTTGCAACAAATTGATGAGCAATGCTTTTTTATAATGCCAAC...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              id                                        description  \\\n",
       "0       FJ217664  Escherichia coli strain 33 Pgm (pgm) gene, par...   \n",
       "1      NM_137148  Drosophila melanogaster phosphoglucomutase 2b ...   \n",
       "2       HUMPGM1A  Human phosphoglucomutase 1 (PGM1) mRNA, comple...   \n",
       "3       AY616650  Escherichia coli strain LW1655F+ Pgi (pgi) gen...   \n",
       "4       DQ363252   Drosophila jambulina Pgi (Pgi) gene, partial cds   \n",
       "5   NM_001289789  Homo sapiens glucose-6-phosphate isomerase (GP...   \n",
       "6       MK461929  Escherichia coli strain 2009_36 plasmid p2009_...   \n",
       "7   XM_001356889  Drosophila pseudoobscura pseudoobscura Eno (Dp...   \n",
       "8       KJ896766  Synthetic construct Homo sapiens clone ccsbBro...   \n",
       "9   AFAS01000074  Escherichia coli PCN009 contig107, whole genom...   \n",
       "10      CK135585  RH55882.3prime RH Drosophila melanogaster norm...   \n",
       "11      KJ905202  Synthetic construct Homo sapiens clone ccsbBro...   \n",
       "12  AFST01000003  Escherichia coli O104:H4 str. C227-11 C227-11_...   \n",
       "13     NM_141785  Drosophila melanogaster mitochondrial aconitas...   \n",
       "14      KJ890616  Synthetic construct Homo sapiens clone ccsbBro...   \n",
       "15        S67363  pdhR=genA, aceE=AceE [Escherichia coli, K-12, ...   \n",
       "16     NM_143411  Drosophila melanogaster pyruvate dehydrogenase...   \n",
       "17      EF576990       Homo sapiens PDHA1 (PDHA1) mRNA, partial cds   \n",
       "18      CP015085  Escherichia coli O25b:H4 chromosome, complete ...   \n",
       "19  NM_001300012  Drosophila melanogaster Succinyl-coenzyme A sy...   \n",
       "20      KJ901911  Synthetic construct Homo sapiens clone ccsbBro...   \n",
       "21      KU997026  Escherichia coli strain VA292 plasmid pDGO100,...   \n",
       "22      LC165517   Drosophila unipectinata CG7430 gene, partial cds   \n",
       "23      KR711450  Synthetic construct Homo sapiens clone CCSBHm_...   \n",
       "24      CP015085  Escherichia coli O25b:H4 chromosome, complete ...   \n",
       "25  NM_001273851  Drosophila melanogaster Ribulose-5-phosphate-3...   \n",
       "26      KJ892020  Synthetic construct Homo sapiens clone ccsbBro...   \n",
       "27      KF678351  Escherichia coli strain ED205 type VI secretio...   \n",
       "28  NM_001299419  Drosophila melanogaster Deoxyribose-phosphate ...   \n",
       "29     NM_015954  Homo sapiens deoxyribose-phosphate aldolase (D...   \n",
       "30      DQ472577  Escherichia coli strain 02-0714 Gnd (gnd) gene...   \n",
       "31      CP012528           Drosophila busckii chromosome X sequence   \n",
       "32      KJ891789  Synthetic construct Homo sapiens clone ccsbBro...   \n",
       "33  AIEW01000031  Escherichia coli DEC1B gecDEC1B.contig.30_1, w...   \n",
       "34      CP012524          Drosophila busckii chromosome 2R sequence   \n",
       "35      KJ905855  Synthetic construct Homo sapiens clone ccsbBro...   \n",
       "\n",
       "        organism                                                seq  \n",
       "0         e.coli  ATGGTGGCCCGGCTGATACCAACGTCACTAAAGTGGTGGAAGACAG...  \n",
       "1     drosophila  CTGATATTCAAGTGTAAGTTTACTAACCAAATTTCTAGTTTGTCAC...  \n",
       "2   homo sapiens  GGGCCGGCCGCCCCTCCGCCAGCCAAGTCCGCCGCTCTGACCCCCG...  \n",
       "3         e.coli  GCTGCCTGGCAGGCACTACAGAAACACTTCGATGAAATGAAAGACG...  \n",
       "4     drosophila  ATTGACAGCGCCAATATGTTCGGCTTCTGGGACTGGGTTGGCGGAC...  \n",
       "5   homo sapiens  AATAGCCCTTACCACCAGCAGACACACATCATCTGTTGTACTTGCT...  \n",
       "6         e.coli  ATTCAGACATCAAAAAACTGTTCGGCGAGGTGGATAAGTCCTCCGG...  \n",
       "7     drosophila  ATGGGGTGGACCACGGATCAGGAACACACACACCTGCTCATCTCCG...  \n",
       "8   homo sapiens  GTTCGTTGCAACAAATTGATGAGCAATGCTTTTTTATAATGCCAAC...  \n",
       "9         e.coli  AGCCGAAGGACATTACCTGGTCATAACGCGGACGCGGTAACGACGC...  \n",
       "10    drosophila  AAATATTTAAATGCTTATGAGTCGGCATTTTTAAAACTACACACAG...  \n",
       "11  homo sapiens  GTTCGTTGCAACAAATTGATGAGCAATGCTTTTTTATAATGCCAAC...  \n",
       "12        e.coli  CCCGATGGTTCTAGGCCCCGTTAGCATCCGTTGCCGCGGCAGGCCG...  \n",
       "13    drosophila  CTTCGACTTCAACAACACGAACTGCTAACCAATAAATCAACAAAAT...  \n",
       "14  homo sapiens  GTTCGTTGCAACAAATTGATGAGCAATGCTTTTTTATAATGCCAAC...  \n",
       "15        e.coli  ATGCACCCGCTTTATGCATGGTTGAAGATGAGTTGCTTAAAAAGAA...  \n",
       "16    drosophila  GCCGGCAACGATAGCGATAGACCCATAGGAAAGCGAACCGCAAACC...  \n",
       "17  homo sapiens  TTCCGTCCGAGAAATTCTCGCAGAGCTTACAGGACGAAAAGGAGGT...  \n",
       "18        e.coli  CTCCACGGAACTGGTGGAAGACCCGGAAGCCATACTGCGCTACGGA...  \n",
       "19    drosophila  CTTCTTGGTCACATTGTGCAGTGATAAGCATCGTGGAACGGCGGCG...  \n",
       "20  homo sapiens  GTTCGTTGCAACAAATTGATGAGCAATGCTTTTTTATAATGCCAAC...  \n",
       "21        e.coli  GATAGGCTCAGATAAACAGACCTTACCCTCGCATCGAGAACCGCTT...  \n",
       "22    drosophila  GCAGCGGCGACAACGTCACCGTGTCCGTGGAGAACGCAAAGACCGG...  \n",
       "23  homo sapiens  GTTCGTTGCAACAAATTGATGAGCAATGCTTTTTTATAATGCCAAC...  \n",
       "24        e.coli  CTCCACGGAACTGGTGGAAGACCCGGAAGCCATACTGCGCTACGGA...  \n",
       "25    drosophila  AGCTCCCCTCTGTGTGACCTTAAGTCAACAAGACGCCACCCGCGAA...  \n",
       "26  homo sapiens  GTTCGTTGCAACAAATTGATGAGCAATGCTTTTTTATAATGCCAAC...  \n",
       "27        e.coli  ATGGCTGACAGTTTCCAGAATGAAGTTCCCACCGCTCGTGTAAATA...  \n",
       "28    drosophila  TGATAAGACTCTCCATGTTCTTCGGTTGAGGTTGAGTAGAAGATTA...  \n",
       "29  homo sapiens  GCGGCGCAGAGGCGGGCGCCTACCAGCCGGCAGCTCCGGAGCTGCC...  \n",
       "30        e.coli  TTCCGGTGGTGAAGAGGGCGCGCTGAAAGGTCCTTCCATTATGCCT...  \n",
       "31    drosophila  CACACACACACACACACACACACACACCCTGCACAACAGCGCCACT...  \n",
       "32  homo sapiens  GTTCGTTGCAACAAATTGATGAGCAATGCTTTTTTATAATGCCAAC...  \n",
       "33        e.coli  CTCTGGGACCAGGAACCTCGAAAATAATAAGCGTAGTAGAATCTGT...  \n",
       "34    drosophila  GGATCTTGCCATCAGGTCCCTCGGTTGGCGTCTGTTTAGTCGCGTT...  \n",
       "35  homo sapiens  GTTCGTTGCAACAAATTGATGAGCAATGCTTTTTTATAATGCCAAC...  "
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.read_sql('SELECT * FROM gene', conn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
